### PR TITLE
[serve] Power of two random choices routing

### LINKED ISSUE #19723: [Serve] Handle should not send tasks on crashed actor
Currently when when an actor crash, serve handle expect that controller will refresh the set of actors. But controller might be down or busy at times. The information about actor status is available through the `ready_ref` we are tracking in `ReplicaSet`. We should be able to instructment serve handle to choose not to send new requests to crashed actors, or exponentially block it for some duration.

### LINKED ISSUE #34280: [serve] Implement backpressure by rejecting queries if `max_concurrent_queries` is exceeded
Currently `max_concurrent_queries` is only respected by the caller. The replica will queue queries indefinitely, as will the HTTP proxy. This causes stability issues if the system is overloaded.

Proposed changes:

- In each replica, if `max_concurrent_queries` is exceeded, raise a `BackpressureException` with a relevant message.
- Support configuring `max_concurrent_queries` in the HTTP options in the config. In the HTTP proxy if `max_concurrent_queries` is exceeded (or a `BackpressureException` is returned by a downstream replica), return a [503 service unavailable](https://www.rfc-editor.org/rfc/rfc7231#section-6.6.4) code.

### LINKED ISSUE #35778: [serve][streaming] Respect `max_concurrent_queries` when using streaming calls

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Adds a feature flag for "power of two random choices" routing that will replace our existing naiive round-robin approach.

The main motivations for this are to:

1. properly enforce `max_concurrent_queries` independent of the number of callers. Note that we can still exceed , but should only be by a constant factor rather than `O(number of callers)`.
2. improve our routing to lower tail latency (particularly when there are variable-length processing times).
3. enforce `max_concurrent_queries` w/ streaming responses (not possible w/ our current implementation).

Benchmark results are [here](https://docs.google.com/spreadsheets/d/1Fha6fDAyB2yFxH7BSGaFvlTzwniJClH9L4APy6nt-6o/edit?usp=sharing), but TL;DR:

- Average & tail latencies decreased and throughput increased over a number of conditions, increasing with variable processing times.
- No-op latencies increased by ~0.5ms (small regression).
- Throughput decreased for condition with a single caller and no-op requests (stressful microbenchmark).

Overall we expect this solution to perform better in real-world scenarios.

Copying from the docstring:

>     """Chooses a replica for each request using the "power of two choices" procedure.
> 
>     Requests are scheduled in FIFO order.
> 
>     When a request comes in, two replicas are chosen randomly as candidates. The queue
>     length of each replicas is requested from it via a control message.
> 
>     The replica responds with two items: (queue_length, accepted). Only replicas that
>     accept the request are considered; between those, the one with the lower queue
>     length is chosen.
> 
>     In the case when neither replica accepts the request (e.g., their queues are full),
>     the procedure is repeated with backoff. This backoff repeats indefinitely until a
>     replica is chosen, so the caller should use timeouts and cancellation to avoid
>     hangs.
> 
>     Each request being scheduled may spawn an independent task that runs the scheduling
>     procedure concurrently. This task will not necessarily satisfy the request that
>     started it (in order to maintain the FIFO order). The total number of tasks is
>     capped at (2 * num_replicas).
>     """

Java and multiplexing support are not included in this PR and will be added in follow-ups.

## Related issue number

Closes https://github.com/ray-project/ray/issues/19723
Closes https://github.com/ray-project/ray/issues/34280
Closes https://github.com/ray-project/ray/issues/35778

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/replica.py [hunk 1]
```diff
@@ -207,6 +207,19 @@ async def initialize_replica():
             # Used to guard `initialize_replica` so that it isn't called twice.
             self._replica_init_lock = asyncio.Lock()
 
+        @ray.method(concurrency_group=HEALTH_CHECK_CONCURRENCY_GROUP)
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/replica.py:210]: Remember to rename the concurrency group.

File: python/ray/serve/_private/router.py [hunk 2]
```diff
@@ -83,62 +125,194 @@ def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
...
-            except StopIteration:
+        self._scheduling_tasks: Set[asyncio.Task] = set()
+        self._pending_assignment_futures: List[asyncio.Future] = []
+
+    @property
+    def num_pending_assignments(self) -> int:
+        return len(self._pending_assignment_futures)
+
+    @property
+    def curr_scheduling_tasks(self) -> int:
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/router.py:178]:
```suggestion
    def num_scheduling_tasks(self) -> int:
```

File: python/ray/serve/_private/router.py [hunk 3]
```diff
@@ -83,62 +125,194 @@ def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
...
-    async def assign_replica(
-        self, query: Query
-    ) -> "ray._raylet.StreamingObjectRefGenerator":
-        replica = None
-        while replica is None:
-            try:
-                replica = next(self._replica_iterator)
-            except StopIteration:
+        self._scheduling_tasks: Set[asyncio.Task] = set()
+        self._pending_assignment_futures: List[asyncio.Future] = []
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/router.py:171]: Could you use a `deque` here, so `append` and `pop(0)` are both constant?

File: python/ray/serve/_private/router.py [hunk 4]
```diff
@@ -83,62 +125,194 @@ def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
...
+        if chosen_replica_id is None:
+            print("No available replicas.")
+            return None
+
+        return self._replicas[chosen_replica_id]
+
+    def schedule_replica(self, replica):
+        while len(self._pending_assignment_futures) > 0:
+            fut = self._pending_assignment_futures.pop(0)
+            if not fut.done():
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/router.py:267]: Add a comment for when `fut.done()` might be `True`.

File: python/ray/serve/_private/router.py [hunk 5]
```diff
@@ -83,62 +125,194 @@ def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
...
+        print(curr_task, "try_schedule entered")
+        keep_scheduling = True
+        while keep_scheduling:
+            async for candidates in self.choose_two_gen():
+                print("GOT CANDIDATES", candidates)
+                replica = await self.select_replica(candidates)
+                print("GOT REPLICA", replica)
+                if replica is not None:
+                    keep_scheduling = self.schedule_replica(replica)
+                    print(curr_task, "Got replica! Keep scheduling?", keep_scheduling)
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/router.py:284]: If you `keep_scheduling`, you should reset the backoff inside the `choose_two_gen` generator. Otherwise, it'll keep using the latest backoff instead of restarting from the shortest backoff.

File: python/ray/serve/tests/test_replica_scheduler.py [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+
+    # Number of tasks should increase when more replicas are available.
+    scheduling_tasks_one_replica = s.curr_scheduling_tasks
+    r2 = FakeReplicaWrapper("r2")
+    r2.set_queue_len_response(0, accepted=False)
+    s.update_replicas([r1, r2])
+    assert s.curr_scheduling_tasks > scheduling_tasks_one_replica
+    assert s.curr_scheduling_tasks == s.max_scheduling_tasks
+
+    # Number of tasks should decrase as the number of pending queries decreases.
```
Comment (shrekris-anyscale) [inline python/ray/serve/tests/test_replica_scheduler.py:277]:
```suggestion
    # Number of tasks should decrease as the number of pending queries decreases.
```

**Review by shrekris-anyscale — COMMENTED** [review]
Nice! At a high level, the logic looks good. Let me know when it's ready for a deeper review.

**Comment by shrekris-anyscale** [comment]
Will this close #19723?

File: python/ray/serve/_private/router.py (hunk 3 above)
Comment (edoakes) [reply python/ray/serve/_private/router.py:184]: done!

**Comment by edoakes** [comment]
Adding duplicate builds w/ the feature flag turned on. Once java & multiplexing are supported it can be turned on by default.

File: python/ray/serve/tests/test_handle.py [hunk 7]
```diff
@@ -225,7 +227,17 @@ def exists(self):
...
+    loop = asyncio.new_event_loop()
+    threading.Thread(
+        daemon=True,
+        target=loop.run_forever,
+    ).start()
+    return loop
+
+
+@pytest.mark.asyncio
+async def test_handle_across_loops(serve_instance):
```
Comment (edoakes) [inline python/ray/serve/tests/test_handle.py:240]: As far as I can tell, this test had a latent race condition where the loop could be closed under some circumstances. I've fixed that by running the loops in a separate thread.

File: python/ray/serve/_private/router.py [hunk 8]
```diff
@@ -83,62 +141,248 @@ def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
...
+        backoff sleep time.
+        """
+        backoff_index = 0
+        while True:
+            # If no replicas are available, wait until `update_replicas` is called.
+            while len(self._replicas) == 0:
                 logger.info(
                     "Tried to assign replica for deployment "
-                    f"{self._deployment_name} but none available.",
+                    f"{self._deployment_name} but none are available.",
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/router.py:249]:
```suggestion
                    f"{self._deployment_name} but none are available. ",
                    "Waiting for new replicas."
```

File: python/ray/serve/_private/router.py [hunk 9]
```diff
@@ -83,62 +141,248 @@ def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
...
+        done, pending = await asyncio.wait(
+            get_queue_len_tasks,
+            timeout=response_deadline_s,
+            return_when=asyncio.ALL_COMPLETED,
         )
+        for task in pending:
+            task.cancel()
+
+        chosen_replica_id = None
+        lowest_queue_len = math.inf
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/router.py:291]: [Nit] Up to you, I prefer this because it removes an import, but I'm okay either way.

```suggestion
        lowest_queue_len = float("inf")
```

File: python/ray/serve/_private/router.py [hunk 10]
```diff
@@ -83,62 +141,248 @@ def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
...
+        """
+        try:
+            while len(self._scheduling_tasks) <= self.target_num_scheduling_tasks:
+                async for candidates in self.choose_two_replicas_with_backoff():
+                    replica = await self.select_from_candidate_replicas(candidates)
+                    if replica is not None:
+                        self.fulfill_next_pending_assignment(replica)
+                        break
+        except Exception:
+            logger.exception("Unexpected error in fulfill_pending_assignments.")
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/router.py:338]: Can you log the exception here?

File: python/ray/serve/_private/router.py [hunk 11]
```diff
@@ -358,11 +596,14 @@ async def assign_replica(self, query: Query) -> ray.ObjectRef:
         and only send a query to available replicas (determined by the
         max_concurrent_quries value.)
         """
+        if query.metadata.is_streaming:
+            raise NotImplementedError("Streaming requires new routing to be enabled.")
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/router.py:600]: Could you explain how to enable new routing in this error message?

**Review by sihanwang41 — COMMENTED** [review]
For multiplexing: 
1. always go through all the replicas which hold the model_Id.
2. if not available(no replica hold the model or all replicas are busy) from step1, go through the `choose_two_replicas_with_backoff` to pick up the replica.

We have to pass the model id information from the request to the candidate selection logic.

File: python/ray/serve/_private/router.py [hunk 12]
```diff
@@ -390,20 +631,28 @@ def __init__(
         controller_handle: ActorHandle,
         deployment_name: str,
         event_loop: asyncio.BaseEventLoop = None,
-        _stream: bool = False,
+        _use_new_routing: bool = False,
     ):
         """Router process incoming queries: assign a replica.
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/router.py:636]: This docstring is a bit awkwardly phrased; could you rephrase it?

**Review by shrekris-anyscale — COMMENTED** [review]
Nice work! I left some more comments.

**Review by GeneDer — APPROVED** [review]
Some comments, but no serious blockers. LGTM!

File: python/ray/serve/_private/router.py (hunk 10 above)
Comment (edoakes) [reply python/ray/serve/_private/router.py:338]: `logger.exception` logs the traceback with it

File: python/ray/serve/_private/router.py [hunk 13]
```diff
@@ -83,62 +141,248 @@ def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
...
+                    extra={"log_to_stderr": False},
+                )
 
-        if replica.is_cross_language:
-            raise RuntimeError(
-                "Streaming is not yet supported for cross-language actors."
+            chosen_ids = random.sample(
+                self._replica_id_set, k=min(2, len(self._replica_id_set))
             )
+            yield [self._replicas[chosen_id] for chosen_id in chosen_ids]
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/router.py:262]: tried this but it requires converting `replicas` to a list each time which is `O(n)`, so prefer to stick w/ sampling from the ID set

File: .buildkite/pipeline.build.yml [hunk 14]
```diff
@@ -214,6 +214,39 @@
...
+  conditions:
+    [
+        "RAY_CI_SERVE_AFFECTED",
+        "RAY_CI_PYTHON_AFFECTED",
+        "RAY_CI_ML_AFFECTED",
+    ]
+  instance_size: large
+  commands:
+    - cleanup() { if [ "${BUILDKITE_PULL_REQUEST}" = "false" ]; then ./ci/build/upload_build_info.sh; fi }; trap cleanup EXIT
+    - TORCH_VERSION=1.9.0 ./ci/env/install-dependencies.sh
```
[the remark this answers is not in the document]
Comment (edoakes) [reply .buildkite/pipeline.build.yml:85]: yeah good point, let's upgrade soon

File: python/ray/serve/_private/router.py [hunk 15]
```diff
@@ -83,62 +141,248 @@ def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
...
-        if new_replica_ids != self._replica_id_set:
-            self._replica_id_set = new_replica_ids
-            logger.info(
-                "Got updated replicas for deployment "
-                f"{self._deployment_name}: {new_replica_ids}.",
-                extra={"log_to_stderr": False},
+        if chosen_replica_id is None:
+            return None
+
+        return self._replicas[chosen_replica_id]
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/router.py:306]: this is a great point; it's possible that the replica is removed from the set as we're waiting for the response. I will address this.

[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/router.py]: I've fixed this and added a test case, see `test_replica_responds_after_being_removed`

File: python/ray/serve/_private/router.py (hunk 11 above)
Comment (edoakes) [reply python/ray/serve/_private/router.py:606]: This is really an internal assertion, should never be hit because we flip this flag on if streaming is enabled

File: python/ray/serve/_private/router.py (hunk 12 above)
Comment (edoakes) [reply python/ray/serve/_private/router.py:642]: on it

File: python/ray/serve/_private/router.py [hunk 16]
```diff
@@ -83,62 +141,248 @@ def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
...
+    Each request being scheduled may spawn an independent task that runs the scheduling
+    procedure concurrently. This task will not necessarily satisfy the request that
+    started it (in order to maintain the FIFO order). The total number of tasks is
+    capped at (2 * num_replicas).
     """
 
-    def __init__(self, deployment_name: str):
+    # The sequence of backoff timeouts to use when all replicas' queues are full.
+    # The last item in the list is the max timeout and will be used repeatedly.
+    backoff_sequence_s = [0, 0.05, 0.1, 0.15, 0.2, 0.5, 1.0]
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/router.py:169]: what do you mean by "fit a line" ?

I chose these somewhat arbitrarily... but they performed well in benchmarks

[the remark this answers is not in the document]
Comment (GeneDer) [reply python/ray/serve/_private/router.py:163]: I mean something like y = min(0.05x, 1), where y is the backoff time and x is the index. So we can just calculate the backoff time directly from the equation and not needing to hard code points. But either way is fine 🙂

File: python/ray/serve/_private/router.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (GeneDer) [reply python/ray/serve/_private/router.py:261]: oh I see, yea that makes sense!

**Review by GeneDer — APPROVED** [review]
:shipit:

File: python/ray/serve/_private/router.py [hunk 17]
```diff
@@ -83,62 +139,255 @@ def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
...
+
+    # Deadline for replicas to respond with their queue length. If the response isn't
+    # received within this deadline, the replica will not be considered.
+    queue_len_response_deadline_s = 0.1
```
Comment (alexeykudinkin) [inline python/ray/serve/_private/router.py:171]: That seems to be quite royal timeout for polling the queue length. Why 100ms?

File: python/ray/serve/_private/router.py [hunk 18]
```diff
@@ -83,62 +139,255 @@ def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
...
+            return_when=asyncio.ALL_COMPLETED,
         )
+        for task in pending:
+            task.cancel()
+
+        chosen_replica_id = None
+        lowest_queue_len = math.inf
+        for task in done:
+            if task.exception() is not None:
+                logger.warning(
```
Comment (alexeykudinkin) [inline python/ray/serve/_private/router.py:297]: This could be pretty noisy (due to timeouts). We should probably consider sampling

File: python/ray/serve/_private/router.py [hunk 19]
```diff
@@ -83,62 +139,255 @@ def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
...
+        finally:
+            self._scheduling_tasks.remove(asyncio.current_task(loop=self._loop))
+
+    def maybe_start_scheduling_tasks(self):
+        """Start scheduling tasks to fulfill pending assignments if necessary.
+
+        Starts tasks so that there is at least one task per pending assignment
+        (respecting the max number of scheduling tasks).
+        """
+        tasks_to_start = (
```
Comment (alexeykudinkin) [inline python/ray/serve/_private/router.py:353]: Can you elaborate why we schedule N tasks instead of just 1 for a single replica assignment?

File: python/ray/serve/_private/router.py [hunk 20]
```diff
@@ -390,20 +636,28 @@ def __init__(
...
         """
         self._event_loop = event_loop
-        if _stream:
-            self._replica_scheduler = RoundRobinStreamingReplicaScheduler(
-                deployment_name
+        if _use_new_routing:
+            self._replica_scheduler = PowerOfTwoChoicesReplicaScheduler(
+                event_loop, deployment_name
+            )
+            logger.info(
```
Comment (alexeykudinkin) [inline python/ray/serve/_private/router.py:651]: nit: We can unify the logging to 1 statement

File: python/ray/serve/_private/common.py [hunk 21]
```diff
@@ -173,9 +173,11 @@ def from_proto(cls, proto: StatusOverviewProto) -> "StatusOverview":
         )
 
 
-HEALTH_CHECK_CONCURRENCY_GROUP = "health_check"
+# Concurrency group used for operations that cannot be blocked by user code
+# (e.g., health checks and fetching queue length).
+CONTROL_PLANE_CONCURRENCY_GROUP = "control_plane"
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/common.py:178]: it's in `serve/_private/common.py`, scoped to serve

File: python/ray/serve/_private/router.py [hunk 22]
```diff
@@ -73,7 +77,59 @@ async def resolve_async_tasks(self):
...
+    """Defines the interface for a scheduler to talk to a replica.
+
+    This is used to abstract away details of Ray actor calls for testing.
+    """
+
+    @property
+    def replica_id(self) -> str:
+        pass
+
+    async def get_queue_len(self) -> Tuple[str, int, bool]:
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/router.py:90]: done

File: python/ray/serve/_private/router.py (hunk 18 above)
Comment (edoakes) [reply python/ray/serve/_private/router.py:297]: timeouts will not be reflected here

File: python/ray/serve/_private/router.py [hunk 23]
```diff
@@ -73,7 +77,59 @@ async def resolve_async_tasks(self):
...
+        self, query: Query
+    ) -> Union[ray.ObjectRef, "ray._raylet.StreamingObjectRefGenerator"]:
+        actor = self.replica_info.actor_handle
+        if query.metadata.is_streaming:
+            obj_ref = actor.handle_request_streaming.options(
+                num_returns="streaming"
+            ).remote(pickle.dumps(query.metadata), *query.args, **query.kwargs)
+        else:
+            _, obj_ref = actor.handle_request.remote(
+                pickle.dumps(query.metadata), *query.args, **query.kwargs
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/router.py:124]: it's very ugly because of the variable number of returns :'(

File: python/ray/serve/_private/router.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/router.py:90]: done

**Comment by edoakes** [comment]
@alexeykudinkin good points re: the timeouts, we can likely tighten them. I've already benchmarked thoroughly with these so don't want to mess w/ them at the moment. I'll try tightening and re-benchmark after getting it in.

File: python/ray/serve/_private/constants.py [hunk 24]
```diff
@@ -203,10 +203,17 @@ class ServeHandleType(str, Enum):
...
 # Feature flag to enable StreamingResponse support.
 # When turned on, *all* HTTP responses will use Ray streaming object refs.
+# Turning this FF on also enables RAY_SERVE_ENABLE_NEW_ROUTING.
 RAY_SERVE_ENABLE_EXPERIMENTAL_STREAMING = (
     os.environ.get("RAY_SERVE_ENABLE_EXPERIMENTAL_STREAMING", "0") == "1"
 )
 
+# Feature flag to enable power of two choices routing.
+RAY_SERVE_ENABLE_NEW_ROUTING = (
+    os.environ.get("RAY_SERVE_ENABLE_NEW_ROUTING", "0") == "1"
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/constants.py:213]: I chose something vague intentionally because otherwise from my experience users will read too much into the details. This should hopefully never need to be seen/touched by users anyways.

File: python/ray/serve/_private/router.py [hunk 25]
```diff
@@ -73,7 +77,59 @@ async def resolve_async_tasks(self):
...
+    @property
+    def replica_id(self) -> str:
+        return self.replica_info.replica_tag
+
+    async def get_queue_len(self) -> Tuple[str, int, bool]:
+        queue_len = (
+            await self.replica_info.actor_handle.get_num_ongoing_requests.remote()
+        )
+        accepted = queue_len < self.replica_info.max_concurrent_queries
+        return self.replica_id, queue_len, accepted
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/router.py:113]: renamed to `get_queue_state`

File: python/ray/serve/_private/router.py [hunk 26]
```diff
@@ -73,7 +88,50 @@ async def resolve_async_tasks(self):
...
+    async def get_queue_len(self) -> Tuple[str, int, bool]:
+        pass
+
+    def send_query(
+        self, query: Query
+    ) -> Union[ray.ObjectRef, "ray._raylet.StreamingObjectRefGenerator"]:
+        pass
+
+
+class ActorReplicaWrapper:
```
[the remark this answers is not in the document]
Comment (shrekris-anyscale) [reply python/ray/serve/_private/router.py:100]: Sounds good.

File: python/ray/serve/_private/router.py (hunk 11 above)
Comment (shrekris-anyscale) [reply python/ray/serve/_private/router.py:610]: Sounds good.

**Review by shrekris-anyscale — APPROVED** [review]
Nice work! This change looks good to me.
